python/tvm/build_module.py

```diff
@@ -68,6 +72,68 @@ def __exit__(self, ptype, value, trace):
 
 BuildConfig.current = BuildConfig()
 
+class DumpIR(object):
+    '''dump ir for each pass'''
+    pass_id = 0
+    decorated = False
+    def __init__(self):
+        self._local_decorated = False
+
+    def decorate(self, func):
+        ''' decorate the pass function'''
+        def dump(*args, **kwargs):
+            '''dump function'''
+            retv = func(*args, **kwargs)
+            if not isinstance(retv, (_stmt.Stmt, container.LoweredFunc, container.Array)):
+                return retv
+            pname = str(DumpIR.pass_id) + "_" + func.func_name + "_ir.cc"
+            with open(pname, "a") as f:
+                out = retv.body if isinstance(retv, container.LoweredFunc) else retv
+                f.write(str(out))
+                if isinstance(retv, container.Array):
+                    for x in retv:
+                        out = x.body if isinstance(x, container.LoweredFunc) else x
+                        f.write("---------", x.name, "\n", str(out), "\n-----------\n")
+                DumpIR.pass_id += 1
+            return retv
+        return dump
+
+    def decorate_irpass(self):
+        for k, v in vars(ir_pass).items():
+            vars(ir_pass)[k] = self.decorate(v) if isinstance(v, types.FunctionType) else v
+        schedule.ScheduleOps = self.decorate(schedule.ScheduleOps)
+
+    def decorate_custompass(self):
+        cfg = BuildConfig.current
+        add_lower_pass = cfg.add_lower_pass if cfg.add_lower_pass else []
+        pass_list = [(x[0], self.decorate(x[1])) for x in add_lower_pass]
+        BuildConfig.current.add_lower_pass = pass_list
+
+    def __enter__(self, pass_inp=None):
+        if DumpIR.decorated is True:
+            return
+        self._old_irpass = vars(ir_pass).items()
```

Comment:
simply have a list of recover functions, which are closure that can get called in exit

Reply:
Example
```python
def decorate_pass(self):
     vset = var(ir_pass)
      for k, v in vset.items():
          def recover():
              vset[k] = v
           self._recover_list.append(recover)

def __exit__(self):
    for f in self._recover_list:
        f()
```